server/pairing/challenge_management.go

```diff
@@ -0,0 +1,247 @@
+package pairing
+
+import (
+	"bytes"
+	"crypto/rand"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+
+	"github.com/btcsuite/btcutil/base58"
+	"github.com/gorilla/sessions"
+	"go.uber.org/zap"
+)
+
+const (
+	// Session names
+	sessionChallenge = "challenge"
+	sessionBlocked   = "blocked"
+)
+
+type ChallengeError struct {
+	Text     string
+	HTTPCode int
+}
+
+func (ce *ChallengeError) Error() string {
+	return fmt.Sprintf("%s : %d", ce.Text, ce.HTTPCode)
+}
+
+func makeCookieStore() (*sessions.CookieStore, error) {
+	auth := make([]byte, 64)
+	_, err := rand.Read(auth)
+	if err != nil {
+		return nil, err
+	}
+
+	enc := make([]byte, 32)
+	_, err = rand.Read(enc)
+	if err != nil {
+		return nil, err
+	}
+
+	return sessions.NewCookieStore(auth, enc), nil
+}
+
+// ChallengeGiver is responsible for generating challenges and checking challenge responses
+type ChallengeGiver struct {
+	cookieStore *sessions.CookieStore
+	encryptor   *PayloadEncryptor
+	logger      *zap.Logger
+	authedIP    net.IP
+}
+
+func NewChallengeGiver(e *PayloadEncryptor, logger *zap.Logger) (*ChallengeGiver, error) {
+	cs, err := makeCookieStore()
+	if err != nil {
+		return nil, err
+	}
+
+	return &ChallengeGiver{
+		cookieStore: cs,
+		encryptor:   e.Renew(),
+		logger:      logger,
+	}, nil
+}
+
+func (cg *ChallengeGiver) getIP(r *http.Request) (net.IP, *ChallengeError) {
+	h, _, err := net.SplitHostPort(r.RemoteAddr)
+	if err != nil {
+		cg.logger.Error("getIP: h, _, err := net.SplitHostPort(r.RemoteAddr)", zap.Error(err), zap.String("r.RemoteAddr", r.RemoteAddr))
+		return nil, &ChallengeError{"error", http.StatusInternalServerError}
+	}
+	return net.ParseIP(h), nil
+}
+
+func (cg *ChallengeGiver) registerClientIP(r *http.Request) *ChallengeError {
+	hIP, err := cg.getIP(r)
```

Comment:
Oh, yeah, that is not clear. A lot of this functionality was in one function, and then I refactored my refactor and he var names moved with the section of code that they were associated with.

So `hIP` means `hostIP` because the IP address is derived from the `r.RemoteAddr` which returns a `host:port` string. We need to split the host and port then parse the host value. That's the background 😆 

But `hostIP` only makes sense in the internal context of `getIP()`, so your suggestion is a good one. Also a good sign is that when rereading this code, I had forgotten myself what `hIP` meant 🙃